File: 0050-powx-n/0050-powx-n.java
class Solution {
    public double myPow(double x, int n) {
        // Handle the case when n is negative
        long N = n;
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }

        double ans = 1.0;

        // Perform exponentiation by squaring
        while (N > 0) {
            // If the current bit is 1 (n is odd)
            if ((N % 2) == 1) {
                ans = ans * x;
            }
            // Square the base for the next bit
            x = x * x;
            // Move to the next bit
            N = N / 2;
        }

        return ans;
    }
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def power(x: float, n: int) -> float:
            if n == 0:  # Base case: any number to power 0 is 1
                return 1
            
            half = power(x, n // 2)
            if n % 2 == 0:
                return half * half
            else:
                return half * half * x

                
        if n < 0: 
            return 1/power(x, abs(n))
        return power(x,n)
        #return ans
            
        